### PR TITLE
ESLint Plugin: Add flanking whitespace and hyphenated range rules

### PR DESCRIPTION
**Pull request by jhnstn**
## Description
While working on translation issues for Gutenberg mobile we noticed that strings with flanking whitespace and hyphens were not ending up translated in the WP Android app. In both cases the strings are being transformed in the translation pipeline which prevents proper look ups when the strings are mounted. 

The source of the flanking whitespace issue was not searched for. Instead after noticing the no collapsable whitespace rule we opted for a new no flanking whitespace  rule since both suffer similar potential issues (we in fact noticed that some locales [included](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/main/res/values-de/strings.xml#L238) trailing whitespace while others [did not](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/main/res/values-en-rGB/strings.xml#L238)). 

The hyphen issue is linked to an Android style lint that suggests using dashes over hyphens in number ranges (See [here](https://android.googlesource.com/platform/tools/base/+/master/lint/libs/lint-checks/src/main/java/com/android/tools/lint/checks/TypographyDetector.java#58) and [here](https://android.googlesource.com/platform/tools/base/+/master/lint/libs/lint-checks/src/main/java/com/android/tools/lint/checks/TypographyDetector.java#152) ). The hint suggests replacing hyphens with dashes in all cases but the rule is only applied to numerical ranges. @fluiddot [recently updated the string transform that enforces this rule](https://github.com/wordpress-mobile/gutenberg-mobile/pull/4484) in Gutenberg Mobile.

Since both issues  only have a noticeable impact in the mobile apps, this PR only applies the rules to `*.native.js` files.

## Testing Instructions

1. Apply the following patch to the root Gutenberg directory 
<details>
 <summary>Test files</summary>

```diff

diff --git a/test-hyphen.js b/test-hyphen.js
new file mode 100644
index 0000000000..966f0935c3
--- /dev/null
+++ b/test-hyphen.js
@@ -0,0 +1,13 @@
+const __ = ( str ) => str;
+
+// Valid cases:
+__( '-10' );
+__( '1–10' ); // Note the use of 'en' dash (U+2013) vs hyphen (U+002D)
+__( '1—10' ); // Note use of 'em' dash (U+2014) vs hyphen (U+002D)
+__( '1\u201310' );
+__( 'Text - with a hyphen - is fine.' );
+
+// Invalid cases:
+__( '1-10' );
+__( '1 - 10' );
+__( '1  -  10' );
diff --git a/test-whitespace.js b/test-whitespace.js
new file mode 100644
index 0000000000..72af663d80
--- /dev/null
+++ b/test-whitespace.js
@@ -0,0 +1,11 @@
+const __ = ( str ) => str;
+
+// Valid cases:
+__( 'ok' );
+
+// Invalid cases:
+__( ' A string with a leading space.' );
+__( 'A string with a trailing space. ' );
+__( ' A string with flanking spaces. ' );
+__( '\tA string with a leading tab.' );
+__( 'A string with an empty newline\n' );

```

</details>

2. Run `npm run lint-js -- test*.js`.
3. Verify that no `@wordpress/i18n-hyphenated-range` or `@wordpress/i18n-no-flanking-whitespace` errors are raised. (Note that `@wordpress/i18n-no-collapsible-whitespace` errors will be raised)
4. Rename each test file with the native file extension i.e. `test-*.native.js`
5. Run `npm run lint-js -- test*.js` again
6. Verify that there are multiple `@wordpress/i18n-hyphenated-range` or `@wordpress/i18n-no-flanking-whitespace` errors raised.
7. Run `npm run lint-js -- test*.js --fix`  
8. Note that no errors are raised
9. Inspect `test-whitespace.native.js` and verify that none of the strings contain flanking whitespace
10. Inspect `test-hyphen.native.js` and verify that the invalid hyphens have been replaced with en dashes

## Types of changes

- Add two new esllint rules
   - `@wordpress/i18n-no-flanking-whitespace`
   - `@wordpress/i18n-hyphenated-range`
- New rules are only applied to `*.native.js` files

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [x] I've updated related schemas if appropriate.

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/i18n-hyphenated-range.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Disallow using a hyphenated numerical ranges in translatable strings (i18n-hyphen-range)
```
Comment (fluiddot) [inline packages/eslint-plugin/docs/rules/i18n-hyphenated-range.md]: Nitpick: Update `i18n-hyphen-range` to use the actual rule's name `i18n-hyphenated-range`.

```suggestion
# Disallow using a hyphenated numerical ranges in translatable strings (i18n-hyphenated-range)
```

File: packages/eslint-plugin/docs/rules/i18n-hyphenated-range.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+```
+
+Examples of **correct** code for this rule:
+
+```js
+__( '-10' );
+__( '1–10' ); // Note the use of 'en' dash (U+2013) vs hyphen (U+002D)
+__( '1—10' ); // Note use of 'em' dash (U+2014) vs hyphen (U+002D)
+__( '1\u201310' );
+__( 'Text - with a hyphen - is fine.' );
```
Comment (fluiddot) [inline packages/eslint-plugin/docs/rules/i18n-hyphenated-range.md]: I noticed that the following string `1 -2` is marked as a lint error, however, is a correct case ([reference](https://android.googlesource.com/platform/tools/base/+/master/lint/libs/lint-checks/src/main/java/com/android/tools/lint/checks/TypographyDetector.java#270)).

**Review by fluiddot — COMMENTED** [review]
Awesome work @jhnstn on adding these new ESLint rules 🥇 .

I haven't finished reviewing the whole PR, but I spotted a case that would be interesting to cover regarding the hyphenated range rule ([reference](https://github.com/WordPress/gutenberg/pull/38225#discussion_r795809581)).

File: packages/eslint-plugin/docs/rules/i18n-hyphenated-range.md (hunk 2 above)
Comment (jhnstn) [reply packages/eslint-plugin/docs/rules/i18n-hyphenated-range.md]: nice, I'll add that 👍

Comment (jhnstn) [reply packages/eslint-plugin/docs/rules/i18n-hyphenated-range.md]: added in https://github.com/WordPress/gutenberg/pull/38225/commits/cdf350f8f18d81183fce150dac1a17e371c42b0e

File: packages/eslint-plugin/rules/__tests__/i18n-hyphenated-range.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+const ruleTester = new RuleTester( {
+	parserOptions: {
+		ecmaVersion: 6,
+	},
+} );
+
+ruleTester.run( 'i18n-hyphenated-range', rule, {
+	valid: [
+		{
+			code: `__( 'Hypenated-words are ok' )`,
```
Comment (fluiddot) [inline packages/eslint-plugin/rules/__tests__/i18n-hyphenated-range.js]: Minor typo:
```suggestion
			code: `__( 'Hyphenated-words are ok' )`,
```

File: packages/eslint-plugin/rules/__tests__/i18n-hyphenated-range.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+		},
+		{
+			code: `__( 'Hyphen - when used in this case - is ok' )`,
+		},
+		{
+			code: `__('en dash – and em dash — are ok')`,
+		},
+		{
+			code: `__( 'en dash – ranges work 1–99 or 2 – 98' )`,
```
Comment (fluiddot) [inline packages/eslint-plugin/rules/__tests__/i18n-hyphenated-range.js:29]: I'm wondering if we could include the em dash case in number ranges, similarly, as it's described [here](https://github.com/WordPress/gutenberg/pull/38225/files#diff-483eb8894c1591bd819c2b989de264853d925ad1cca409cc9cee4005008f86feR21).

File: packages/eslint-plugin/rules/i18n-hyphenated-range.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+const EN_DASH = '–';
+const HYPHEN_IN_RANGE = /(\d\s+-\s+\d)|(\d-\d)/g;
+
+function replaceHyphenWithEnDash( string ) {
+	if ( string.match( HYPHEN_IN_RANGE ) ) {
+		return string.replace( '-', EN_DASH );
+	}
+}
+
+// see eslint-plugin-wpcalypso.
```
Comment (fluiddot) [inline packages/eslint-plugin/rules/i18n-hyphenated-range.js]: Not sure what means this comment is, is this rule somehow related to `eslint-plugin-wpcalypso`?

File: packages/eslint-plugin/rules/i18n-hyphenated-range.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	};
+}
+
+module.exports = {
+	meta: {
+		type: 'problem',
+		schema: [],
+		messages: {
+			foundHyphen:
+				'Use dashes (en or em) in place of hypens for numeric ranges.',
```
Comment (fluiddot) [inline packages/eslint-plugin/rules/i18n-hyphenated-range.js]: Minor typo:
```suggestion
				'Use dashes (en or em) in place of hyphens for numeric ranges.',
```

File: packages/eslint-plugin/rules/i18n-hyphenated-range.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+					) {
+						fixes.push(
+							fixer.replaceTextRange(
+								[ quasi.start, quasi.end ],
+								replaceHyphenWithEnDash( quasi.value.raw )
+							)
+						);
+					}
+					return fixes;
+				}, [] );
```
Comment (fluiddot) [inline packages/eslint-plugin/rules/i18n-hyphenated-range.js:38]: I noticed that the fix on template literals doesn't produce the expected result:

Input: ```__( `1-2` )```
Output: ```__( 1–2 )```
**Expected output:** ```__( `1–2` )```

However, looks like this issue can also be reproduced with the rule `i18n-ellipsis`:

Input: __( `Some text...` )
Output: __( Some text… )
**Expected output:** __( `Some text…` )

Not sure if you based the implementation on that rule but at least these two produce the same issue.

<hr>

Probably this issue is related to the text range provided in line 32.

File: packages/eslint-plugin/rules/i18n-hyphenated-range.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+						arg,
+						replaceHyphenWithEnDash( arg.raw )
+					),
+				];
+
+			case 'BinaryExpression':
+				return [
+					...makeFixerFunction( arg.left )( fixer ),
+					...makeFixerFunction( arg.right )( fixer ),
+				];
```
Comment (fluiddot) [inline packages/eslint-plugin/rules/i18n-hyphenated-range.js:52]: I verified this works in most cases, however, I spotted a specific case where it produces an unexpected result:
Input: `__( '1-2' + 'Text...' )`
Output: `__( '1–2' + undefined )`
**Expected output:** `__( '1-2' + 'Text…' )`

Similarly, the rule `i18n-ellipsis` has the same issue.

File: packages/eslint-plugin/rules/i18n-no-flanking-whitespace.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+					) {
+						fixes.push(
+							fixer.replaceTextRange(
+								[ quasi.start, quasi.end ],
+								`'${ quasi.value.value.trim() }'`
+							)
+						);
+					}
+					return fixes;
+				}, [] );
```
Comment (fluiddot) [inline packages/eslint-plugin/rules/i18n-no-flanking-whitespace.js]: This rule doesn't give an error when using template literals (example: ```__( ` Text ` )```). I think it might be related to the way to extract the value from `quasi`, looks like we should use `quasi.value.raw` instead of `quasi.value.value`.

_Referece: https://astexplorer.net/_
Screenshot 2022-02-01 at 11 45 14

**Review by fluiddot — COMMENTED** [review]
I tested both rules and they work as expected, although I identified some cases, which I'm not sure if we should consider as edge cases, that might be interesting to investigate further. @jhnstn let me know what you think about this, I'd agree with omitting them for now, in order to merge the rules, and tackle them in the future via new issues.

On the other hand, I spotted a couple of minor typos that would be great to address before merging.

File: packages/eslint-plugin/rules/i18n-hyphenated-range.js (hunk 5 above)
Comment (jhnstn) [reply packages/eslint-plugin/rules/i18n-hyphenated-range.js]: This was from copying the ellipsis rule as a template, I'll remove it.

File: packages/eslint-plugin/docs/rules/i18n-hyphenated-range.md (hunk 2 above)
Comment (jhnstn) [reply packages/eslint-plugin/docs/rules/i18n-hyphenated-range.md]: fixed in https://github.com/WordPress/gutenberg/pull/38225/commits/a30012e88d68ed06e6fd71183cb78e8635bcf587

File: packages/eslint-plugin/rules/i18n-hyphenated-range.js (hunk 7 above)
Comment (jhnstn) [reply packages/eslint-plugin/rules/i18n-hyphenated-range.js:38]: I tried adjusting the fixer and nothing seemed to work quite right. I also noticed that the ellipsis fixer also has issues with template literals. 
I opted to just remove the fixer for template literals, the rule will still catch them but for now the author will have to fix them manually.

Updated here https://github.com/WordPress/gutenberg/pull/38225/commits/7abccc0cc64d01d2db1034965feb6b7e50a0216a

File: packages/eslint-plugin/rules/i18n-hyphenated-range.js (hunk 8 above)
Comment (jhnstn) [reply packages/eslint-plugin/rules/i18n-hyphenated-range.js:52]: Fixed in https://github.com/WordPress/gutenberg/pull/38225/commits/7e795d0dad70d6472dd8c42e95edf67bf084d5c2

I'm guessing `i18n-ellipsis` can be fixed in a similar manner.

File: packages/eslint-plugin/rules/i18n-no-flanking-whitespace.js (hunk 9 above)
Comment (jhnstn) [reply packages/eslint-plugin/rules/i18n-no-flanking-whitespace.js]: Thanks for linking to the AST, that was really helpful Carlos!

I didn't dig too deep but the few things I tried didn't seem to work, or at least I was unable to get the unit tests to work well with any updates. Not that we absolutely need a unit test for this case. I figure since the rule will still catch the case, it's safe to remove the template literal fixer for now.

File: packages/eslint-plugin/rules/__tests__/i18n-hyphenated-range.js (hunk 4 above)
Comment (jhnstn) [reply packages/eslint-plugin/rules/__tests__/i18n-hyphenated-range.js:29]: Personally I'm convinced that the test you linked along with these `en` focused tests are enough to verify that the rule allows non-hyphened ranges. 
I'm open to add another test case if you think it's important.

File: packages/eslint-plugin/rules/i18n-hyphenated-range.js (hunk 7 above)
Comment (fluiddot) [reply packages/eslint-plugin/rules/i18n-hyphenated-range.js:38]: > I opted to just remove the fixer for template literals, the rule will still catch them but for now the author will have to fix them manually.
> 
> Updated here [7abccc0](https://github.com/WordPress/gutenberg/commit/7abccc0cc64d01d2db1034965feb6b7e50a0216a)

I see the template literals fixer was removed from `18n-no-flanking-whitespace` rule but not this one. From your comment, I understand that we would remove it from both rules, right?

File: packages/eslint-plugin/rules/i18n-hyphenated-range.js (hunk 8 above)
Comment (fluiddot) [reply packages/eslint-plugin/rules/i18n-hyphenated-range.js:52]: Great, I confirm that changes from https://github.com/WordPress/gutenberg/commit/7e795d0dad70d6472dd8c42e95edf67bf084d5c2 fixed the issue, thanks for addressing it 🙇 .

File: packages/eslint-plugin/rules/i18n-no-flanking-whitespace.js [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+	return ( fixer ) => {
+		switch ( arg.type ) {
+			case 'Literal':
+				return [ fixer.replaceText( arg, `'${ arg.value.trim() }'` ) ];
+
+			case 'BinaryExpression':
+				return [
+					...makeFixerFunction( arg.left )( fixer ),
+					...makeFixerFunction( arg.right )( fixer ),
+				];
```
Comment (fluiddot) [inline packages/eslint-plugin/rules/i18n-no-flanking-whitespace.js:28]: Nitpick: I noticed the following minor issue.

Input: `__( 'Text is ' + 'separated ' )`
Output: `__( 'Text is' + 'separated' )`
**Expected output:** `__( 'Text is ' + 'separated' )`

In this case, the first text shouldn't be trimmed as its ending whitespace is valid.

**Review by fluiddot — APPROVED** [review]
I spotted a couple of minor issues that we could consider skipping if you wish, apart from that, the PR LGTM 🎊 ! Nice work @jhnstn 🏅 !

File: packages/eslint-plugin/rules/i18n-hyphenated-range.js (hunk 7 above)
Comment (jhnstn) [reply packages/eslint-plugin/rules/i18n-hyphenated-range.js:38]: Ah correct, I'll go ahead and remove the fixer here too. TBH the fixers are "nice to haves" so don't want to spend too much time on them.

File: packages/eslint-plugin/rules/i18n-no-flanking-whitespace.js (hunk 10 above)
Comment (jhnstn) [reply packages/eslint-plugin/rules/i18n-no-flanking-whitespace.js:28]: Good catch! Again fixing this fixer might get complicated with concatenated strings. I'm thinking the simplest thing is to have the fixer only apply to the simplest case. Thoughts?  

Hmm I'll have to check to see if `__(  'This is a ' + 'valid string' )` fails since the trailing whitespace in the first string segment should be valid. Joining strings in the `__` function doesn't not seem like a good idea but something like that might trip up the lint rule.

File: packages/eslint-plugin/rules/i18n-hyphenated-range.js (hunk 7 above)
Comment (fluiddot) [reply packages/eslint-plugin/rules/i18n-hyphenated-range.js:38]: > Ah correct, I'll go ahead and remove the fixer here too.

Just in case it was overlooked, I noticed in the [merge commit](https://github.com/WordPress/gutenberg/commit/1a8971ba51d1f69b718bf2f77261184b43950217#diff-86670005a144398f0d3c02df4d5e838e32707224a8495b13c97fce05b086feef) that the fixer wasn't removed.

> TBH the fixers are "nice to haves" so don't want to spend too much time on them.

I agree, the important part is the rule itself, I'd consider fixers as optional/nice-to-have 👍 .

File: packages/eslint-plugin/rules/i18n-no-flanking-whitespace.js (hunk 10 above)
Comment (fluiddot) [reply packages/eslint-plugin/rules/i18n-no-flanking-whitespace.js:28]: > Good catch! Again fixing this fixer might get complicated with concatenated strings. I'm thinking the simplest thing is to have the fixer only apply to the simplest case. Thoughts?

Yep, the fixer turns out quite complex in this case. I'm wondering if we could check for each argument if the fixer should be applied 🤔. In any case, as we did for other issues with fixers, we could drop them as they're not critical to have.

> Hmm I'll have to check to see if __( 'This is a ' + 'valid string' ) fails since the trailing whitespace in the first string segment should be valid. Joining strings in the __ function doesn't not seem like a good idea but something like that might trip up the lint rule.

I did test the `__( 'This is a ' + 'valid string' )` case and it was marking it as valid, I guess this result makes sense , as when the rule is evaluated for this case, the argument (`argumentString`) is processed as a single string (i.e. `'This is a valid string'`) which is a valid string.
